[AMDGPU] Fix hidden kernarg preload count inconsistency

kerbowa opened the pull request:
It is possible that the number of hidden arguments that are selected to be preloaded in AMDGPULowerKernel arguments and isel can differ. This isn't an issue with explicit arguments since isel can lower the argument correctly either way, but with hidden arguments we may have alignment issues if we try to load these hidden arguments that were added to the kernel signature.

The reason for the mismatch is that isel reserves an extra synthetic user SGPR for module LDS.

Instead of teaching lowerFormalArguments how to handle these properly it makes more sense and is less expensive to fix the mismatch and assert if we ever run into this issue again. We should never be trying to lower these in the normal way.

In a future change we probably want to revise how we track "synthetic" user SGPRs and unify the handling in GCNUserSGPRUsageInfo. Sometimes synthetic SGPRSs are considered user SGPRs and sometimes they are not. Until then this patch resolves the inconsistency, fixes the bug, and is otherwise a NFC.

kerbowa committed:
[AMDGPU] Fix hidden kernarg preload count inconsistency

It is possible that the number of hidden arguments that are selected to
be preloaded in AMDGPULowerKernel arguments and isel can differ. This
isn't an issue with explicit arguments since isel can lower the argument
correctly either way, but with hidden arguments we may have alignment
issues if we try to load these hidden arguments that were added to the
kernel signature.

The reason for the mismatch is that isel reserves an extra synthetic
user SGPR for module LDS.

Instead of teaching lowerFormalArguments how to handle these properly it
makes more sense and is less expensive to fix the mismatch and assert if
we ever run into this issue again. We should never be trying to lower
these in the normal way.

In a future change we probably want to revise how we track "synthetic"
user SGPRs and unify the handling in GCNUserSGPRUsageInfo. Sometimes
synthetic SGPRSs are considered user SGPRs and sometimes they are not.
Until then this patch resolves the inconsistency, fixes the bug, and is
otherwise a NFC.

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -2999,6 +2999,14 @@ SDValue SITargetLowering::LowerFormalArguments(
+#ifndef NDEBUG
+        if (Arg.isOrigArg()) {
+          Argument *OrigArg = Fn.getArg(Arg.getOrigArgIndex());
+          assert(!OrigArg->hasAttribute("amdgpu-hidden-argument") &&
+                 "Hidden arguments should be preloaded");
```
No error on valid IR can be an assert. This should report a proper error (via DiagnosticInfo)

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULowerKernelArguments.cpp:
```diff
@@ -143,17 +143,17 @@ class PreloadKernelArgInfo {
-
-    NumFreeUserSGPRs = MaxUserSGPRs - UserSGPRInfo.getNumUsedUserSGPRs();
+    NumFreeUserSGPRs =
+        UserSGPRInfo.getNumFreeUserSGPRs() - 1 /* Synthetic SGPRs*/;
```
Not sure what synthetic means here

kerbowa replied:
We call `LDSKernelId` a synthetic SGPR which is allocated after normal user SGPRs but before system SGPRs. I just realized this doesn't actually matter for kernels since it should never be used there.

arsenm reviewed (commented):
> It is possible that the number of hidden arguments that are selected to be preloaded in AMDGPULowerKernel arguments and isel can differ.

This should not happen. The introduction of arguments to preload should be precise

kerbowa committed:
Diagnostic error and move LDSKernelId.

Use diagnostic error. Unify preload kernarg SGPR accounting and move
LDSKernelId to UserSGPRInfo so that synthetic SGPRs can be properly
tracked when calculating the number of available registers for preloads.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -520,6 +520,17 @@ bool AMDGPUCallLowering::lowerFormalArgumentsKernel(
 
   // TODO: Align down to dword alignment and extract bits for extending loads.
   for (auto &Arg : F.args()) {
+    // Hidden arguments that are in the kernel signature must be preloded to
```
Typo 'preloded'

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -520,6 +520,17 @@ bool AMDGPUCallLowering::lowerFormalArgumentsKernel(
+    if (Arg.hasAttribute("amdgpu-hidden-argument")) {
+      DiagnosticInfoUnsupported NonPreloadHiddenArg(
+          *Arg.getParent(),
+          "Hidden argument in kernel signature was not preloaded");
```
Start error messages with lowercase

arsenm commented on llvm/lib/Target/AMDGPU/GCNSubtarget.cpp:
```diff
@@ -748,6 +748,10 @@ GCNUserSGPRUsageInfo::GCNUserSGPRUsageInfo(const Function &F,
     FlatScratchInit = true;
   }
 
+  if (!AMDGPU::isGraphics(CC) && !IsKernel &&
+      !F.hasFnAttribute("amdgpu-no-lds-kernel-id"))
+    LDSKernelId = true;
+
```
The kernel ID has to exist for all callable functions, !graphics will be wrong for amdgpu_gfx. Should be not entry function cc?

kerbowa replied:
I've just removed the refactoring for LDSKernelID since it's not relevant for kernels, but FYI this just copied the existing logic from `SIMachineFunction.cpp`, so if it's incorrect we should update it there.

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -3020,6 +3021,21 @@ SDValue SITargetLowering::LowerFormalArguments(
+          if (OrigArg->hasAttribute("amdgpu-hidden-argument")) {
+            DiagnosticInfoUnsupported NonPreloadHiddenArg(
+                *OrigArg->getParent(),
+                "Hidden argument in kernel signature was not preloaded",
```
Start messages with lowercase

arsenm commented on llvm/lib/Target/AMDGPU/GCNSubtarget.h:
```diff
@@ -1692,11 +1694,12 @@ class GCNUserSGPRUsageInfo {
     FlatScratchInitID = 6,
-    PrivateSegmentSizeID = 7
+    PrivateSegmentSizeID = 7,
+    LDSKernelIdID = 8
```
Comment this isn't really a user SGPR? Maybe should rename the enum

arsenm commented on llvm/test/CodeGen/AMDGPU/invalid-hidden-kernarg-in-kernel-signature.ll:
```diff
@@ -0,0 +1,17 @@
+; RUN: not llc -global-isel=1 -mtriple=amdgcn--amdhsa -mcpu=gfx940 -start-after=amdgpu-lower-kernel-arguments < %s 2>&1 | FileCheck -check-prefix=ERROR %s
+; RUN: not llc -global-isel=0 -mtriple=amdgcn--amdhsa -mcpu=gfx940 -start-after=amdgpu-lower-kernel-arguments < %s 2>&1 | FileCheck -check-prefix=ERROR %s
```
Shouldn't need the -start-after

arsenm commented on llvm/lib/Target/AMDGPU/GCNSubtarget.cpp:
```diff
@@ -771,6 +775,9 @@ GCNUserSGPRUsageInfo::GCNUserSGPRUsageInfo(const Function &F,
   if (hasPrivateSegmentSize())
     NumUsedUserSGPRs += getNumUserSGPRForField(PrivateSegmentSizeID);
+
+  if (hasLDSKernelId())
+    NumSyntheticSGPRs += getNumUserSGPRForField(LDSKernelIdID);
```
I'm not sure I understand why anything is changing with the kernel ID. Is this just changing bookkeeping to split out the kernel's user SGPRs. vs. the implicit SGPRs in a function?

arsenm commented on llvm/lib/Target/AMDGPU/GCNSubtarget.h:
```diff
@@ -1676,11 +1676,13 @@ class GCNUserSGPRUsageInfo {
   unsigned getNumUsedUserSGPRs() const { return NumUsedUserSGPRs; }
 
-  unsigned getNumFreeUserSGPRs();
+  unsigned getNumFreeKernargPreloadSGPRs();
```
Document

arsenm commented on llvm/lib/Target/AMDGPU/GCNSubtarget.h:
```diff
@@ -1676,11 +1676,13 @@ class GCNUserSGPRUsageInfo {
 
   bool hasPrivateSegmentSize() const { return PrivateSegmentSize; }
 
+  bool hasLDSKernelId() const { return LDSKernelId; }
```
I don't know why this is defined in GCNSubtarget.h, it's not related to the subtarget

kerbowa committed:
temp

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -520,6 +520,16 @@ bool AMDGPUCallLowering::lowerFormalArgumentsKernel(
 
   // TODO: Align down to dword alignment and extract bits for extending loads.
   for (auto &Arg : F.args()) {
+    // Hidden arguments that are in the kernel signature must be preloaded to
+    // user SGPRs, or loaded via the implicit_arg ptr. Print a diagnostic error
+    // if a hidden argument is in the argument list and is not preloaded.
+    if (Arg.hasAttribute("amdgpu-hidden-argument")) {
+      DiagnosticInfoUnsupported NonPreloadHiddenArg(
+          *Arg.getParent(),
+          "hidden argument in kernel signature was not preloaded");
+      F.getContext().diagnose(NonPreloadHiddenArg);
+    }
```
This is an unconditonal error, and untested. Currently the IR kernel argument lowering is treated as optional, both selectors are supposed to fully handle kernel lowering independently with -amdgpu-ir-lower-kernel-arguments=0.

I also don't understand the second part of the comment. If they're loaded through the implicit pointer, they weren't preloaded?

kerbowa replied:
We had decided a while back that hidden arguments should not always be in the kernel signature. They will only be present if they are preloaded into user SGPRs. This diagnostic is saying we should never see the attribute in gisel since preloading kernargs isn't implemented there yet, but I'm going to add that support next.

If a hidden argument in the signature and has the attribute `amdgpu-hidden-argument` uses of that argument must use the preloaded SGPR. There is nothing stopping other instances of the same argument from being accessed via the implicit_arg ptr in a callee for example.

We could easily make it so that isel can handle these even if they are not preloaded, but since it is not intended and it indicates a mismatch between isel and `AMDGPULowerKernelArguments` we error out instead.

arsenm replied:
I would expect globalisel to be at feature parity, and hit the fallback for unsupported features, not hard error

arsenm commented on llvm/test/CodeGen/AMDGPU/invalid-hidden-kernarg-in-kernel-signature.ll:
```diff
@@ -0,0 +1,17 @@
+; RUN: not llc -global-isel=1 -mtriple=amdgcn--amdhsa -mcpu=gfx940 < %s 2>&1 | FileCheck -check-prefix=ERROR %s
+; RUN: not llc -global-isel=0 -mtriple=amdgcn--amdhsa -mcpu=gfx940 < %s 2>&1 | FileCheck -check-prefix=ERROR %s
+
```
Also test with the ir lowering disabled

arsenm commented on llvm/test/CodeGen/AMDGPU/invalid-hidden-kernarg-in-kernel-signature.ll:
```diff
@@ -0,0 +1,17 @@
+; RUN: not llc -global-isel=1 -mtriple=amdgcn--amdhsa -mcpu=gfx940 < %s 2>&1 | FileCheck -check-prefix=ERROR %s
+; RUN: not llc -global-isel=0 -mtriple=amdgcn--amdhsa -mcpu=gfx940 < %s 2>&1 | FileCheck -check-prefix=ERROR %s
```
Use 942

kerbowa committed:
Address comments.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -520,6 +520,16 @@ bool AMDGPUCallLowering::lowerFormalArgumentsKernel(
+      DiagnosticInfoUnsupported NonPreloadHiddenArg(
+          *Arg.getParent(),
+          "hidden argument in kernel signature was not preloaded");
+      F.getContext().diagnose(NonPreloadHiddenArg);
```
return false and comment (with debug print) this is just because it's not implemented in globalisel

kerbowa committed:
Abort in gisel.
